src/hotspot/share/services/mallocLimit.cpp

```diff
@@ -0,0 +1,234 @@
+static const char* const MODE_FATAL = "fatal";
+
+static const char* mode_to_name(MallocLimitMode m) {
+  switch (m) {
+  case MallocLimitMode::trigger_fatal: return MODE_FATAL;
+  case MallocLimitMode::trigger_oom: return MODE_OOM;
+  default: ShouldNotReachHere();
+  };
+  return nullptr;
+}
+
+class ParserHelper {
+  // Start, end of parsed string.
+  const char* const _s;
+  const char* const _end;
+  // Current parse position.
+  const char* _p;
+
+public:
+  ParserHelper(const char* s) : _s(s), _end(s + strlen(s)), _p(s) {}
+
+  bool eof() const { return _p >= _end; }
+
+  // Check if string at position matches a malloclimit_mode_t.
+  // Advance position on match.
+  bool match_mode_flag(MallocLimitMode* out) {
+    if (eof()) {
+      return false;
+    }
+    if (strncasecmp(_p, MODE_OOM, strlen(MODE_OOM)) == 0) {
+      *out = MallocLimitMode::trigger_oom;
+      _p += 3;
+      return true;
+    } else if (strncasecmp(_p, MODE_FATAL, strlen(MODE_FATAL)) == 0) {
+      *out = MallocLimitMode::trigger_fatal;
+      _p += 5;
+      return true;
+    }
+    return false;
+  }
+
+  // Check if string at position matches a category name.
+  // Advances position on match.
+  bool match_category(MEMFLAGS* out) {
+    if (eof()) {
+      return false;
+    }
+    const char* end = strchr(_p, ':');
+    if (end == nullptr) {
+      end = _end;
+    }
+    stringStream ss;
+    ss.print("%.*s", (int)(end - _p), _p);
+    MEMFLAGS f = NMTUtil::string_to_flag(ss.base());
+    if (f != mtNone) {
+      *out = f;
+      _p = end;
+      return true;
+    }
+    return false;
+  }
+
+  // Check if string at position matches a memory size (e.g. "100", "100g" etc).
+  // Advances position on match.
+  bool match_size(size_t* out) {
+    if (!eof()) {
+      char* remainder = nullptr;
+      if (parse_integer<size_t>(_p, &remainder, out)) {
+        assert(remainder > _p && remainder <= _end, "sanity");
+        _p = remainder;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // Check if char at pos matches c; return true and advance pos if so.
+  bool match_char(char c) {
+    if (!eof() && (*_p) == c) {
+      _p ++;
+      return true;
+    }
+    return false;
+  }
+};
+
+MallocLimitSet::MallocLimitSet() {
+  reset();
+}
+
+void MallocLimitSet::set_global_limit(size_t s, MallocLimitMode flag) {
+  _glob.sz = s; _glob.mode = flag;
+}
+
+void MallocLimitSet::set_category_limit(MEMFLAGS f, size_t s, MallocLimitMode flag) {
+  const int i = NMTUtil::flag_to_index(f);
+  _cat[i].sz = s; _cat[i].mode = flag;
+}
+
+void MallocLimitSet::reset() {
+  set_global_limit(0, MallocLimitMode::trigger_fatal);
+  _glob.sz = 0; _glob.mode = MallocLimitMode::trigger_fatal;
+  for (int i = 0; i < mt_number_of_types; i++) {
+    set_category_limit(NMTUtil::index_to_flag(i), 0, MallocLimitMode::trigger_fatal);
+  }
+}
+
+void MallocLimitSet::print_on(outputStream* st) const {
+  static const char* flagnames[] = { MODE_FATAL, MODE_OOM };
```

Comment:
This still requires us to be careful and match the order in

enum class MallocLimitMode {
  trigger_fatal = 0,
  trigger_oom   = 1
};

so I like my proposal better, but at least we're not hardcoding it here, so it's an improvement.